examples/cmd/main.go

```diff
@@ -81,50 +86,111 @@ type Ghsync struct {
 func (c *Ghsync) Execute(args []string) error {
 	return c.ExecuteBody(
 		log.New(log.Fields{"org": c.Name}),
-		func(httpClient *http.Client, downloader *github.Downloader) error {
-			repos, err := listRepositories(context.TODO(), httpClient, c.Name, c.NoForks)
+		func(logger log.Logger, dp *DownloadersPool) error {
+			err := c.downloadOrg(logger, dp)
 			if err != nil {
 				return err
 			}
 
-			err = downloader.DownloadOrganization(context.TODO(), c.Name, c.Version)
+			repos, err := c.listRepos(logger, dp)
 			if err != nil {
-				return fmt.Errorf("failed to download organization %v: %v", c.Name, err)
+				return err
 			}
 
-			for _, repo := range repos {
-				err = downloader.DownloadRepository(context.TODO(), c.Name, repo, c.Version)
-				if err != nil {
-					return fmt.Errorf("failed to download repository %v/%v: %v", c.Name, repo, err)
-				}
-			}
+			return c.downloadRepos(logger, dp, repos)
+		})
+}
 
-			return nil
+func (c *Ghsync) downloadOrg(logger log.Logger, dp *DownloadersPool) error {
+	err := dp.WithDownloader(func(d *DownloaderPoolItem) error {
+		logger.Infof("downloading organization")
+		return d.DownloadOrganization(context.TODO(), c.Name, c.Version)
+	})
 
-		})
+	if err != nil {
+		return fmt.Errorf("failed to download organization %v: %v", c.Name, err)
+	}
+
+	logger.Infof("finished downloading organization")
+	return nil
 }
 
-type bodyFunc = func(httpClient *http.Client, downloader *github.Downloader) error
+func (c *Ghsync) listRepos(logger log.Logger, dp *DownloadersPool) ([]string, error) {
+	var repos []string
+	err := dp.WithDownloader(func(d *DownloaderPoolItem) error {
+		var err error
+		logger.Infof("listing repositories")
+		repos, err = d.ListRepositories(context.TODO(), c.Name, c.NoForks)
+		return err
+	})
 
-func (c *DownloaderCmd) ExecuteBody(logger log.Logger, fn bodyFunc) error {
-	client := oauth2.NewClient(context.TODO(), oauth2.StaticTokenSource(
-		&oauth2.Token{AccessToken: c.Token},
-	))
+	if err != nil {
+		return nil, fmt.Errorf("failed to list repositories for organization %v: %v", c.Name, err)
+	}
+
+	logger.Infof("found %d repositories", len(repos))
+	return repos, nil
+}
 
-	if c.LogHTTP {
-		setLogTransport(client, logger)
+func (c *Ghsync) downloadRepos(logger log.Logger, dp *DownloadersPool, repos []string) error {
+	var wg sync.WaitGroup
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	errCh := make(chan error, dp.Size)
+
+	for _, repo := range repos {
+		wg.Add(1)
+		go func(logger log.Logger, r string) {
+			defer wg.Done()
+
+			err := dp.WithDownloader(func(d *DownloaderPoolItem) error {
+				logger.Infof("downloading repository")
+				return d.DownloadRepository(ctx, c.Name, r, c.Version)
+			})
+
+			if ctx.Err() != nil {
+				logger.Warningf("stopped downloading due to context cancellation")
+				return
+			}
+
+			if err != nil {
+				logger.Errorf(err, "error while downloading repository")
```

Comment:
same comment as before. duplicated error logging

Reply:
sorry, but I don't understand this 😅 
I mean by running it and manually provoking the error I don't understand when it's duplicated.

Reply:
okay, I'll fix the wrapping, but I think that it's ok to still log the error here too so that it appears in the _right place_.

Reply:
Don't get me wrong I don't mind keeping this log message. I'm just trying to understand what makes this error different from all the others.
Like `failed to download organization` or `failed to list repositories for organization` would be shown only once but `error while downloading repository` twice...

Reply:
Ah ok! I get your point now. The only difference is that the other errors are followed immediately by the exit of the program, while in this case, other logs could be printed. But I don't like this inconsistency as well, and IMO this difference is not enough to justify it. I'll just remove it for now and open a separate issue to discuss and decide what approach we may want to apply. Maybe it's better to also log the other errors, as now if we decide to redirect only logs I think if I'm not wrong that we lose these errors.